### PR TITLE
Migrate Edge calls for Worker to FastAPI part 3 - Jobs routes

### PR DESCRIPTION
**Pull request by jscheffl**
Follow-up PR as incremental part of https://github.com/apache/airflow/pull/44311 and #44330

Note: Only the last commit is the relevant change, the first 4 commits are from https://github.com/apache/airflow/pull/44311 and #44330

To prepare EdgeWorker to be independent of AIP-44 Internal API, this PR is the third step in adding/migrating to FastAPI. The calls to "Jobs" API to (1) fetch a job and (2) report job state are now real REST API calls, not using internal API.

I would separate the other internal API calls to follow-up PRs as this is already quite large. Especially cause for ongoing Airflow 2.10 Connexion API + Swagger manually need to be generated whereas the main workstream for Airflow 3 uses FastAPI.

### CODE REVIEW

File: providers/src/airflow/providers/edge/worker_api/datamodels.py
Comment (kaxil) [inline providers/src/airflow/providers/edge/worker_api/datamodels.py]: While we have a comment about using `str` because of connexion, we have it as int here? Can we make it one or the other in the entire edge provider?

File: providers/src/airflow/providers/edge/worker_api/datamodels.py
Comment (kaxil) [inline providers/src/airflow/providers/edge/worker_api/datamodels.py]: Providers are now Py 3.9+ , so we can make this `list[str] | None`

Comment (kaxil) [reply providers/src/airflow/providers/edge/worker_api/datamodels.py]: and ideally move to `Annotated` type and move description there

File: providers/src/airflow/providers/edge/worker_api/routes/_v2_routes.py
Comment (kaxil) [inline providers/src/airflow/providers/edge/worker_api/routes/_v2_routes.py:286]:
```py
queues = body.get("queues")
````

File: providers/src/airflow/providers/edge/worker_api/routes/_v2_routes.py
Comment (kaxil) [inline providers/src/airflow/providers/edge/worker_api/routes/_v2_routes.py:287]:
```py
free_concurrency = body.get("free_concurrency", 1)
````

File: providers/src/airflow/providers/edge/worker_api/routes/_v2_routes.py
Comment (kaxil) [inline providers/src/airflow/providers/edge/worker_api/routes/_v2_routes.py:279]: Isn't the return type deterministic: `EdgeJobFetched | None`?

File: providers/src/airflow/providers/edge/worker_api/routes/_v2_routes.py
Comment (kaxil) [inline providers/src/airflow/providers/edge/worker_api/routes/_v2_routes.py:301]: its str here why int above! https://github.com/apache/airflow/pull/44433/commits/a7df2861cf97ae1c8e14b766bbb7b8d687058acc#r1862844010

File: providers/src/airflow/providers/edge/worker_api/routes/jobs.py
Comment (kaxil) [inline providers/src/airflow/providers/edge/worker_api/routes/jobs.py]:
```py
update(EdgeJobModel).where(
        EdgeJobModel.dag_id == dag_id,
        EdgeJobModel.task_id == task_id,
        EdgeJobModel.run_id == run_id,
        EdgeJobModel.map_index == map_index,
        EdgeJobModel.try_number == try_number,
    )
).values(state=state, last_update=timezone.utcnow())
result = session.execute(query)
log.info("State updated: %s row(s) affected", result.rowcount)
```

**Review by kaxil — APPROVED** [review]

**Comment by potiuk** [comment]
I will wait for 1 and 2 to be merged :)

**Comment by jscheffl** [comment]
> I will wait for 1 and 2 to be merged :)

As internal API is now "gone" I will have a hard time getting #1+#2 merged. As there are some merge conflicts as well I'll focus on PR#3 + #4...

But anyway let me resolve the conflicts and integrate the feedback from Kaxil then it might be cleaner already.

**Comment by potiuk** [comment]
> UPDATE: Okay now I realize that the internal API is gone but all the previous functions are moved to provider package... need to catch-up coming home...

:eyes:

File: providers/src/airflow/providers/edge/worker_api/datamodels.py
Comment (jscheffl) [reply providers/src/airflow/providers/edge/worker_api/datamodels.py]: I have this on my bucket list to check. Maybe the problem has been removed the last days when switching to Pydantic 2.10.2. It was a problem when pytest was runnign in backcompat and Pydantic could not load because not able to evaluate the type.

File: providers/src/airflow/providers/edge/worker_api/routes/_v2_routes.py
Comment (jscheffl) [reply providers/src/airflow/providers/edge/worker_api/routes/_v2_routes.py:279]: That would be the type in FastAPI. But here we are in the "compatability leyer" which uses the fucntions of FastAPI and wraps into the same endpoint for Connexion. Connexion gets the JSON model as dump, not a pydantic object.

File: providers/src/airflow/providers/edge/worker_api/datamodels.py
Comment (jscheffl) [reply providers/src/airflow/providers/edge/worker_api/datamodels.py]: Just Flask has a parsing problem and therefore the Airflow 2.10 REST implementation in Connexion uses "string" for the map-index as there can be negative values. But I convert this to int() before going into the logic.

The primary implementation is FastAPI and this can correctly treat negative integer values. So just in the openapi specs of connexion and the `_v2_routes.py` the map_index is treated as string to make it working. Otherwise FastAPI and Pydantic models should be consistent.

File: providers/src/airflow/providers/edge/worker_api/routes/_v2_routes.py
Comment (jscheffl) [reply providers/src/airflow/providers/edge/worker_api/routes/_v2_routes.py:301]: As answer above. Actually in all internals I want to have `int` just the connexion wrapper entry uses `str`to work-around the flask negative int value limitation

File: providers/src/airflow/providers/edge/worker_api/routes/jobs.py
Comment (jscheffl) [reply providers/src/airflow/providers/edge/worker_api/routes/jobs.py]: Ah, that is a good thing to optimize. Should have jumped into my eyes earlier!

File: providers/src/airflow/providers/edge/worker_api/datamodels.py
Comment (jscheffl) [reply providers/src/airflow/providers/edge/worker_api/datamodels.py]: Okay, pushed the commit to clean this up and seems the upgrade of pydantic fixed it. Added this in the last commit.

**Comment by jscheffl** [comment]
> I will wait for 1 and 2 to be merged :)

You just merged 2 after I merged 1... so "ball is in your field" now :-D

File: providers/src/airflow/providers/edge/worker_api/routes/jobs.py
Comment (potiuk) [reply providers/src/airflow/providers/edge/worker_api/routes/jobs.py]: That's a strange suggestion copilot

**Review by potiuk — COMMENTED** [review]
NICE... Way easier to review once 1 and 2 merged :)

**Review by potiuk — APPROVED** [review]

File: providers/src/airflow/providers/edge/worker_api/routes/jobs.py
Comment (jscheffl) [reply providers/src/airflow/providers/edge/worker_api/routes/jobs.py]: @potiuk I also first thought so... first time... then reading again... I really had a copy&past error being made and the description was really for a different case.
But here the worker reports the queues and capacity he supports which will be a filter criteria for the service to select the next job.

Comment (potiuk) [reply providers/src/airflow/providers/edge/worker_api/routes/jobs.py]: AHA.. Cool.

Comment (kaxil) [reply providers/src/airflow/providers/edge/worker_api/routes/jobs.py]: Copilot is getting smarter - nice

Comment (potiuk) [reply providers/src/airflow/providers/edge/worker_api/routes/jobs.py]: Indeed. Almost un-human with this check 😱

Comment (potiuk) [reply providers/src/airflow/providers/edge/worker_api/routes/jobs.py]: Wait... It's not human :)

**Pull request merged**
